### PR TITLE
fix: fix RedirectUri sonarcloud refactor code-coverage unit-tests

### PR DESCRIPTION
**Pull request by jmunozherbas**
1.- Fixed, all sonar issues
2.- inproved, code coverage percent with unit test ( methods with logic)

### CODE REVIEW

File: jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java [hunk 1]
```diff
@@ -103,11 +107,23 @@ public void addResponseParameter(String key, String value) {
...
     public String getResponseParameter(@NotNull String key) {
         return responseParameters.get(key);
     }
 
+    public int getResponseParamentersSize() {
+        int size = 0;
+        if (responseParameters != null) {
+            size = responseParameters.size();
+        }
+        return size;
```
Comment (Milton-Ch) [inline jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java]:
```suggestion
        return responseParameters != null ? responseParameters.size() : 0;
```

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+        String queryResult = redirectUri.getQueryString();
+
+        assertNoEmptyQueryString(queryResult, RESPONSE, 1);
+
+    }
+
+    @Test()
+    public void toString_withResponseModeFormPostJwtAndNoParams_NoEmpty() {
+
```
Comment (Milton-Ch) [inline jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java]: Remove empty line?

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+        RedirectUri redirectUri = getRedirectUriTemplateGetQueryString(null, null, null, null);
+        redirectUri.addResponseParameter(AuthorizeResponseParam.CODE, "123");
+        redirectUri.addResponseParameter(AuthorizeResponseParam.EXPIRES_IN, "3000");
+        redirectUri.addResponseParameter(AuthorizeResponseParam.ACCESS_TOKEN, "tk123");
+        String queryResult = redirectUri.getQueryString();
+
+        assertNoEmptyQueryString(queryResult, AuthorizeResponseParam.CODE, 3);
+    }
+
+    @Test()
```
Comment (Milton-Ch) [inline jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java]: You can remove parentheses, they are not required here.

File: jans-auth-server/common/pom.xml [hunk 4]
```diff
@@ -219,5 +219,15 @@
...
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-testng</artifactId>
+            <version>0.4.8</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
```
Comment (Milton-Ch) [inline jans-auth-server/common/pom.xml]: Worths checking if we can include it as part of a lower level, perhaps bom or similar level.

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+        RedirectUri redirectUri = getRedirectUriTemplateGetQueryString(null, null, null, null);
+        redirectUri.parseQueryString(null);
+        assertEquals(redirectUri.getResponseParamentersSize(), 0);
+    }
+
+    @Test
+    public void parseQueryString_oneParam_size1() throws CryptoProviderException {
+        RedirectUri redirectUri = getRedirectUriTemplateGetQueryString(null, null, null, null);
+        redirectUri.parseQueryString(AuthorizeResponseParam.EXPIRES_IN + "=");
+        assertEquals(redirectUri.getResponseParamentersSize(), 1);
```
Comment (Milton-Ch) [inline jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java:72]: I'd suggest to add more asserts here in order to validate right params were gotten.

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    @Test
+    public void parseQueryString_threeParams_size3() throws CryptoProviderException {
+        RedirectUri redirectUri = getRedirectUriTemplateGetQueryString(null, null, null, null);
+        redirectUri.parseQueryString(AuthorizeResponseParam.CODE + "=Pcode&" + AuthorizeResponseParam.EXPIRES_IN + "=3000&" + AuthorizeResponseParam.ACCESS_TOKEN + "=tk123");
+        assertEquals(redirectUri.getResponseParamentersSize(), 3);
```
Comment (Milton-Ch) [inline jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java:86]: Similar to my previous comment, length doesn't define that the whole method was processed properly.
I'd suggest to extract params and compare them with the real params.

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+
+    @Test()
+    public void getQueryString_responseModeFormPostJWT_NoEmpty() throws CryptoProviderException {
+        RedirectUri redirectUri = getRedirectUriTemplateGetQueryString(ResponseMode.FORM_POST_JWT, null, null, null);
+        AbstractCryptoProvider cryptoProvider = mock(AbstractCryptoProvider.class);
+        when(cryptoProvider.sign(anyString(), anyString(), anyString(), Mockito.any())).thenReturn("12345");
+        redirectUri.setCryptoProvider(cryptoProvider);
+        String queryResult = redirectUri.getQueryString();
+        //No empty Result
+        assertTrue(queryResult.length() > 0);
```
Comment (Milton-Ch) [inline jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java:129]: I'd suggest to add `verify` sentences at the end for CryptoProvider as well. Verify that certain methods were called with specific values.

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+        when(cryptoProvider.getPublicKey(anyString(), Mockito.any(), Mockito.any())).thenReturn(null);
+        redirectUri.setCryptoProvider(cryptoProvider);
+
+        String queryResult = redirectUri.getQueryString();
+        assertEquals(queryResult, "");
+    }
+
+    @Test()
+    public void getQueryString_withEncriptionAlgorithmRSANoSignatureAlgorithm_responseEncoded() throws UnsupportedEncodingException, CryptoProviderException, InvalidJweException {
+        RedirectUri redirectUri = getRedirectUriTemplateGetQueryString(ResponseMode.JWT, KeyEncryptionAlgorithm.RSA1_5, BlockEncryptionAlgorithm.A256GCM, null);
```
Comment (Milton-Ch) [inline jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java]: Minor: try to keep line length limit.

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+            public byte[] getEncoded() {
+                return new byte[0];
+            }
+
+            @Override
+            public BigInteger getModulus() {
+                return null;
+            }
+        };
+    }
```
Comment (Milton-Ch) [inline jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java]: I'm guessing that you can replace this by using `mock` of RSAPublicKey instead.

File: jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java [hunk 10]
```diff
@@ -356,25 +369,20 @@ private String getJweResponse(String nestedJws) throws Exception {
...
-            if (publicKey != null) {
-                JweEncrypter jweEncrypter = new JweEncrypterImpl(keyEncryptionAlgorithm, blockEncryptionAlgorithm, publicKey);
-                jwe = jweEncrypter.encrypt(jwe);
-            } else {
+            if (publicKey == null) {
                 throw new InvalidJweException("The public key is not valid");
             }
+
+            JweEncrypter jweEncrypter = new JweEncrypterImpl(keyEncryptionAlgorithm, blockEncryptionAlgorithm, publicKey);
+            jwe = jwtForEncrypt.encryptJwe(jwe, jweEncrypter);
```
Comment (Milton-Ch) [inline jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java]: Why are you using jwtForEncrypt? do you really require to extract it outside? I mean, why cannot be jwtEncrypter?

File: jans-auth-server/common/pom.xml (hunk 4 above)
Comment (yuriyz) [reply jans-auth-server/common/pom.xml]: I do agree, version should go to at least parent pom.

File: jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java [hunk 11]
```diff
@@ -67,12 +69,14 @@
     private String nestedSharedSecret;
     private JSONObject nestedJsonWebKeys;
     private AbstractCryptoProvider cryptoProvider;
+    private Jwe jwtForEncrypt;
```
Comment (yuriyz) [inline jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java:63]: We have to strive to keep variables as narrow in scope as possible. This one went to class level while it has to be local to method without obvious reason. In short, lets drop it.

File: jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java [hunk 12]
```diff
@@ -399,52 +407,57 @@ private void appendFragmentSymbol(StringBuilder sb) {
...
+                appendQuerySymbol(sb);
             }
+            sb.append(getQueryString());
+        } else {
+            appendDeafultToString(sb);
         }
         return sb.toString();
     }
+
+    private void appendDeafultToString(StringBuilder sb) {
```
Comment (yuriyz) [inline jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java:455]: Typo: Deafult -> Default

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+
+    @Test()
+    public void toString_withResponseModeFormPostJwtAndNoParams_NoEmpty() {
+
+        RedirectUri redirectUri = getRedirectUriTemplateToString();
+        redirectUri.setResponseMode(ResponseMode.FORM_POST_JWT);
+        redirectUri.setCryptoProvider(cryptoProvider);
+
+        assertTrue(redirectUri.toString().length() > 0, "Is Empty");
```
Comment (yuriyz) [inline jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java]: toString() is important test. In each `toString()` test we should see equals assertion
```
assertEquals("expected value", redirectUri.toString());
```
Also when I look into that method after your refactoring I am not sure logic of the method is not broken. If you cover it with tests correctly (with equals) we will know it exactly. It is main goal of writing these tests. It is not to get good coverage number but really be sure that logic inside is exactly what we expect it to be.

I would suggest:
1. take old version of class
2. cover it with equals test
3. refactor it
4. Run tests from step 2. In this way you will be sure that refactoring does not broke anything.

File: jans-auth-server/model/src/main/java/io/jans/as/model/jwe/Jwe.java [hunk 14]
```diff
@@ -49,6 +49,10 @@ public static Jwe parse(String encodedJwe, PrivateKey privateKey, byte[] sharedS
         return jwe;
     }
 
+    public Jwe encryptJwe(Jwe jwe, JweEncrypter jweEncrypter) throws InvalidJweException {
```
Comment (yuriyz) [inline jans-auth-server/model/src/main/java/io/jans/as/model/jwe/Jwe.java:51]: Lets drop this method. Not clear why do we need it here. If we need higher level of abstraction for encryption then lets add it. Calling jwe drom jweEncrypter or jweEncrypted from jwe does not solve anything.

File: jans-auth-server/common/pom.xml (hunk 4 above)
Comment (jmunozherbas) [reply jans-auth-server/common/pom.xml]: I have removed the version from jans-common pom,xml, and put it on parent pom.xml.
I have included the dependency in jans-bom pom.xml too

File: jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java (hunk 11 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java:63]: The variable was deleted.

File: jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java (hunk 1 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java]: got it, refactored

File: jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java (hunk 10 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java]: jwtEncrypter was deleted from class.

File: jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java (hunk 12 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/main/java/io/jans/as/common/util/RedirectUri.java:455]: Got it, thanks

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java (hunk 5 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java:72]: More Asserts, was added

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java (hunk 6 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java:86]: More Asserts, was added

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java (hunk 3 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java]: Got it, thanks

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java (hunk 8 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java]: Got it.

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java (hunk 9 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java]: I have replaced it, with a method better resumed.

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java (hunk 2 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java]: Got it, thanks.

File: jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java (hunk 13 above)
Comment (jmunozherbas) [reply jans-auth-server/common/src/test/java/io/jans/as/common/RedirectUriTest.java]: I have refactored tests, including more asserts and assertEquals in toString tests.
according your suggestion, i have tested previous and last version of the method, the results was successfully.

**Pull request closed without merging**